### PR TITLE
Add shuffleAll feature to song audio player

### PR DESCRIPTION
**Pull request by clwluvw**
Added shuffle all feature for song audio player. Related to #4640 

I'm not good in designing so I didn't create an icon for shuffle and I used repeat icon. Can someone add it and push it to this branch?

Signed-off-by: Seena Fallah <seeenafallah@gmail.com>

### CODE REVIEW

**Comment by damnfella** [comment]
No offense, but Telegram isn't supposed to be used as a music player. If you do add the shuffle feature, it would only mean that Telegram is supporting the mass distribution of pirated music.

(This isn't a legal advice, but it certainly can be interpreted this way by those who want to destroy Telegram by any means possible)

**Comment by clwluvw** [comment]
> (This isn't a legal advice, but it certainly can be used as an argument by those who want to destroy Telegram legally or damage its reputation)

@damnfella Are you kidding me? Telegram for iOS already has this feature!

**Comment by RememberTheAir** [comment]
> P.S. I am aware that Telegram Android (beta) has the shuffle feature hidden.

It's not hidden at all. The Android app having it is enough to legitimize a feature request on tdesktop

> If you do add the shuffle feature, it would only mean that Telegram is supporting user's enjoyment of mass distributed pirated music on its platform.

Uhhh... how does the "shuffle" feature makes Telegram more or less supportive to pirated music sharing? We already have a fully-fledged music player in the mobile apps, and almost a complete one on tdesktop, I guess this is enough of a statement already

**Comment by damnfella** [comment]
Don't get me wrong. I would love a feature like this. But I think it shouldn't really exist.

@RememberTheAir 
> It's not hidden at all. 
I mean, You have to tap on the reorder icon to see the option in the dropdown.

> Uhhh... how does the "shuffle" feature makes Telegram more or less supportive of pirated music sharing? 
1. It would "allow enjoyment on its own platform", which can also translate to it's "encouragement" in a way.
2. Cloud based messaging.
3. 50 MB limit on music streaming (more or less).

Of course, these are just my thoughts on it. I may be very wrong here. 🤷

**Comment by Aokromes** [comment]
> I'm not good in designing so I didn't create an icon for shuffle and I used repeat icon. Can someone add it and push it to this branch?

get android icon for that :)

**Comment by IlyaMrz** [comment]
> No offense, but Telegram isn't supposed to be used as a music player. 
nonsense
> certainly can be used as an argument by those who want to destroy Telegram legally
it can't

**Comment by IlyaMrz** [comment]
> No offense, but Telegram isn't supposed to be used as a music player. 

nonsense
> certainly can be used as an argument by those who want to destroy Telegram legally

it can't

**Comment by vgdh** [comment]
It will be cool! A want shuffle mode too ))

**Comment by clwluvw** [comment]
Hi @john-preston, This PR needs https://github.com/desktop-app/lib_base/pull/18

File: .gitignore [hunk 1]
```diff
@@ -18,6 +18,7 @@ Release/
 *.xcodeproj
 ipch/
 .vs/
+.vscode/
```
Comment (ilya-fedin) [inline .gitignore:21]: It is better to use `.git/info/exclude` for this

**Comment by xtrahigh** [comment]
> Telegram isn't supposed to be used as a music player
But in the 11 oct 2017 update we have new music player features and I think Telegram developers better provide this simple but really useful feature. Especially if the Android app has this feature.

**Comment by Nirzak** [comment]
People who think this feature will encourage piracy, well, giving anyone to play the media files shared by them, it's not encouraging piracy. All social platforms give that facility, and telegram just improving it by giving it a music player experience. So, I think it will be a nice feature, and telegram should improve the music and video player experience. Piracy totally depends on the users who are sharing those files.

**Comment by ilya-fedin** [comment]
AFAIK the blocker is the lack of local playlist, so the feature will work really weird across restart

**Comment by deulizealand** [comment]
> No offense, but Telegram isn't supposed to be used as a music player. If you do add the shuffle feature, it would only mean that Telegram is supporting user's enjoyment of mass distributed pirated music on its platform.
> 
> (This isn't a legal advice, but it certainly can be used as an argument by those who want to destroy Telegram legally or damage its reputation)
> 
> P.S. I am aware that Telegram Android (beta) has the shuffle feature hidden.

its make sense.

**Comment by john-preston** [comment]
This won't work as expected :( because it shuffles only a small loaded part of the playlist, not the whole playlist + adding icons to the existing player design looks really poor on small window sizes.

Also this implementation relies on random shuffling base::flat_set of ids, but this breaks base::flat_set contract of order preserving (base::flat_set is a sorted container of unique elements) and may lead to unexpected results in different places.

I hope a redesign of player (somewhen in the future, I don't know when) can bring such option here, but it requires additional work with API to be able to load non-yet-loaded parts of the playlist if a randomly chosen track happens to be in such part.

**Pull request closed without merging**